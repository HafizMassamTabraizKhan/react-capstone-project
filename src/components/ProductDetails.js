import React, { useEffect } from 'react';
import { useParams, Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchProductById,
  removeSelectedProduct,
} from '../redux/products/productsSlice';

const ProductDetails = () => {
  const { productId } = useParams();
  const dispatch = useDispatch();
  const product = useSelector((state) => state.products.selectedProduct);
  const status = useSelector((state) => state.products.status);
  const error = useSelector((state) => state.products.error);

  useEffect(() => {
    dispatch(fetchProductById(productId));

    return () => {
      dispatch(removeSelectedProduct());
    };
  }, [productId, dispatch]);

  if (status === 'loading') {
    return <div>Loading...</div>;
  }

  if (status === 'failed') {
    return (
      <div>
        Error:
        {error}
      </div>
    );
  }

  return (
    <div className="product-details" data-testid="productDetailPage">
      <Link className="links" to="/">
        <h2>Back to Home</h2>
      </Link>
      {product ? (
        <div className="details-container">
          <div className="headings">
            <h2>{product.title}</h2>
            <div className="rating">
              <h3>
                Price: $
                {product.price}
              </h3>
            </div>
          </div>
          <div className="product-main">
            <img src={product.image} alt="" />
            <div className="right-section">
              <p>{product.description}</p>
              <div className="product-categories">
                <span key={product.category}><b>{product.category}</b></span>
              </div>
              <button type="button" className="btn btn-primary">Add to Cart</button>
            </div>
          </div>
        </div>
      ) : (
        <div>Product not found.</div>
      )}
    </div>
  );
};

export default ProductDetails;
