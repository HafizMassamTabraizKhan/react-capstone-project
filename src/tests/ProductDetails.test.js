import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import PropTypes from 'prop-types';
import store from '../redux/store';
import ProductDetails from '../components/ProductDetails';

class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    console.error(error, errorInfo);
  }

  render() {
    const { hasError } = this.state;
    const { children } = this.props;

    if (hasError) {
      return <div>Error occurred. Please check the console for details.</div>;
    }

    return children;
  }
}

describe('ProductDetails', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <ErrorBoundary>
              <ProductDetails />
            </ErrorBoundary>
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

ErrorBoundary.propTypes = {
  children: PropTypes.node.isRequired,
};
