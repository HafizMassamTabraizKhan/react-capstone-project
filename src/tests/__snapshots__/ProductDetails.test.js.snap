// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails renders correctly 1`] = `
<div
  className="product-details"
  data-testid="productDetailPage"
>
  <a
    className="links"
    href="/"
    onClick={[Function]}
  >
    <h2>
      Back to Home
    </h2>
  </a>
  <div>
    Product not found.
  </div>
</div>
`;
